### DISCUSSION TITLE
Is it possible to type configuration settings (array or object)

### DISCUSSION THREAD

**Discussion by sketchbuch** [microsoft/vscode-discussions#864] [Extension Development QnA] [answered]
When using array/object settings, is there anyway to provide a Typescript type so that users know what should be entered in these configuration arrays/objects when editing the settings json?

**Comment by tamuratak** [comment, marked as answer]
You can use [JSON Schema](https://json-schema.org/overview/what-is-jsonschema).

- https://code.visualstudio.com/api/references/contribution-points#Configuration-property-schema

The `properties` and `items` properties are what you are looking for. It would be useful to see  the `package.json` files of the built-in extensions:

https://github.com/microsoft/vscode/blob/e2387821ab562b6d7bafbda7c7c387d4c18e9d7b/extensions/git/package.json#L2691C26-L2715

```json
        "git.showActionButton": {
          "type": "object",
          "additionalProperties": false,
          "description": "%config.showActionButton%",
          "properties": {
            "commit": {
              "type": "boolean",
              "description": "%config.showActionButton.commit%"
            },
            "publish": {
              "type": "boolean",
              "description": "%config.showActionButton.publish%"
            },
            "sync": {
              "type": "boolean",
              "description": "%config.showActionButton.sync%"
            }
          },
          "default": {
            "commit": true,
            "publish": true,
            "sync": true
          },
          "scope": "resource"
        },
```

https://github.com/microsoft/vscode/blob/e2387821ab562b6d7bafbda7c7c387d4c18e9d7b/extensions/git/package.json#L2244-L2266

```json
        "git.checkoutType": {
          "type": "array",
          "items": {
            "type": "string",
            "enum": [
              "local",
              "tags",
              "remote"
            ],
            "enumDescriptions": [
              "%config.checkoutType.local%",
              "%config.checkoutType.tags%",
              "%config.checkoutType.remote%"
            ]
          },
          "uniqueItems": true,
          "markdownDescription": "%config.checkoutType%",
          "default": [
            "local",
            "remote",
            "tags"
          ]
        },
```

**Reply by sketchbuch** [reply]
Thanks a lot!

**Reply by sketchbuch** [reply]
After trying, this is not what I want,

I want a hint to people editing an array/object config in setings.json as to what they should enter. properties just adds additional config properties to a parent config option, but thanks anyway
